### PR TITLE
Add higher-dim projection operators, child_size function for AMR

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

These operators can be used to project data across AMR levels. They are essentially moved from `MortarHelpers`.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Domain/Structure/ChildSize.hpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "Domain/Structure/SegmentId.hpp"
+#include "NumericalAlgorithms/Spectral/Projection.hpp"
```
Comment (kidder) [inline src/Domain/Structure/ChildSize.hpp:7]: please change the include of Matrix.hpp to a forward declaration in Projection.hpp

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Projection.cpp [hunk 2]
```diff
@@ -311,3 +306,14 @@ SPECTRE_TEST_CASE("Unit.Numerical.Spectral.Projection.h.element_to_mortar",
     }
   }
 }
+
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Numerical.Spectral.Projection.p.mortar_to_element",
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_Projection.cpp]: Update the test name for the merged tests.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by kidder** [comment]
also fix the clang-tidy error

**Comment by nilsvu** [comment]
Pushed the forward-declaration as extra commit and squashed in the test name change and the clang-tidy fix. Cheers!

**Review by wthrowe — DISMISSED** [review]

**Review by kidder — DISMISSED** [review]

**Comment by nilsvu** [comment]
Added a missing `<ostream>` include

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**